llama.cpp

```diff
@@ -1342,29 +1342,15 @@ static bool llama_model_load(
     }
 }
 
-// evaluate the transformer
-//
-//   - lctx:         llama context
-//   - tokens:       new batch of tokens to process
-//   - n_past:       the context size so far
-//   - n_threads:    number of threads to use
-//   - cgraph_fname: filename of the exported computation graph
-//
-static bool llama_eval_internal(
-        llama_context &  lctx,
-    const llama_token *  tokens,
-            const int    n_tokens,
-            const int    n_past,
-            const int    n_threads,
-            const char * cgraph_fname) {
-
-    // enforce that the first token is BOS
-    if (n_past == 0 && tokens[0] != llama_token_bos()) {
-        fprintf(stderr, "%s: first token must be BOS\n", __func__);
-        return false;
-    }
-
-    const int64_t t_start_us = ggml_time_us();
+static bool llama_eval_internal_tensor(
+            llama_context& lctx,
+            ggml_context* ctx0,
+            ggml_tensor* inpL,
+            const int   n_tokens,
+            const int   n_past,
+            const int   n_threads,
+            const char * cgraph_fname,
+            const int64_t t_start_us) {
```

Comment:
This is not ideal - the `ctx0` should still be created here instead of passing it from outside.

Better way to extend the API would be like this:

```c++
static bool llama_eval_internal(
        llama_context &  lctx,
    const llama_token *  tokens,
            const int    n_tokens,
          const float *  embd,
            const int    n_past,
            const int    n_threads,
            const char * cgraph_fname) {
```

And assert that `(!tokens && embd) || (tokens && !embd)`
Depending on what is provided - either pick the rows from the model or straight up use `embd`